lms/security.py

```diff
@@ -78,47 +79,67 @@ def forget(self, request):
         return self._policy_getter(request).forget(request)
 
 
-class LTISecurityPolicy:
+class _LTIUserBasedSecurityPolicy:
+    """Base class for `LTIUser` based security policies."""
+
     @classmethod
-    def authenticated_userid(cls, request):
-        if request.lti_user is None:
-            return None
+    def get_lti_user(cls, request) -> LTIUser:  # pragma: no cover
+        """
+        Get and LTIUser to use as the authenticated user in this policy.
+
+        How this LTIUser is constructed is up to the subclasses.
+        """
+        raise NotImplementedError()
```

Comment:
I reckon inheritance is the right tool here but it might not be evident because there's only one class in this PR, in a follow if using composition we'd need:

```python
class LTILaunchesPolicy: # New class
## This is all the same to LTISecurityPolicy
  def __init__(self):
        self._helper = LTIUserAuthenticationHelper()

    def permits(self, *args, **kwargs):
        # _permits() is already a compositional helper, just a function not a class.
        return _permits(*args, **kwargs)

    # There's probably no need to inherit or use a helper for remember() and 
    # forget() if they don't actually do anything, and it's nice to be able to see
    # easily right here in LTISecurityPolicy that it doesn't support remember() and
    # forget(). You will have to duplicate these lines in each security policy but :shrug:
    def remember(self, *args, **kwargs):
         pass

    def forget(self, *args, **kwargs):
        pass

    # authenticated_userid() and identity() seem to be where LTIUserAuthenticationHelper
    # provides its real value.
    def authenticated_userid(self, request):
        lti_user = self._get_lti_user(request)
        return self._helper.authenticated_userid(lti_user, request)

    def identity(self, request):
        lti_user = self._get_lti_user(request)
        return self._helper.identity(lti_user, request)

# This is different
     def _get_lti_user(self, request):
        ...
```

We could probably still trim some of that repetition but again, I think here inheritance is the answer, we know now that all the classes we are going to implement are not going to extend any of the other methods (we are refactoring the existing ones into this shape)  .

If keep adding policies and having to modify any other methods we can always refactor again to use composition instead.

Reply:
> It would probably make unit tests easier to write as you can unit test helper functions or helper classes and then separately unit test the security policies that use those helpers, patching or passing in mocks in place of the helpers. With subclassing it's not possible for tests to patch or inject the base class in order to test the subclasses so you can't cleanly break things into separate units for testing separately.

I disagree with this, we only need to test the method that the sub classes define



We could go the composing route but based on an `lti_getter` and use the same class for all of them


```python
def LTIUserSecurityPolicy:
  def __init__(self, get_lti_user):
    self._get_lti_user = get_lti_user
    ...
   def identity(self, request):
     lti_user = self._get_lti_user(request)
     return self._helper.identity(lti_user, request)
  
```

and on our SecurityPolicy do something like


```python
if path.starswith("/lti_launches`): 
  return LTIUserSecurityPolicy(launches_lti_user)
```

Reply:
Added a version of the "injected _get_lti_user" version in this commit:

https://github.com/hypothesis/lms/pull/4394/commits/a6faf7cab3f6bd7dcc51a5f325d55a2667b392b8


I prefer this approach but it means that instead of having a subclass (or a separate class with composition) per policy we'll have a function/callable for each and that's configured on the top level "get_policy".